### PR TITLE
Better parity with macOS XML plist parsing

### PR DESCRIPTION
**Pull request by korylprince**
This PR intends to bring this library in better alignment with macOS's plist parsing (e.g. plutil, PlistBuddy). This stems from some discussion on the MacAdmins Slack where MDM commands successfully parsed by this library were rejected by macOS's plist parser.

There's two commits:

* add a test suite of valid and invalid plists as determined by `plutil -lint`
  * if plutil is present in `$PATH`, the test will use the actual command to compare parse errors. Otherwise it will use a pre-baked list of invalid plists
  * this should make it easier to add similar test cases in the future if other issues arise
* clean up the XML parser by:
  * only allowing specific tags on for loops in parsing `<plist>`, `<dict>`, and `<array>` tags, so that extraneous tags or chardata returns an error like plutil does
  * verifying that all start tags are escaped with end tags, like plutil does

It should be noted that it's not possible to gain full parity with plutil. For example an malformed DOCTYPE like

```<!DOCTYPE PUBLIC "-//Apple//DTD PLIST>```

 will be happily parsed by plutil, but Go's XML parser will go to EOF looking for the missing `"`, with no control to stop it.

However, this PR should bring us to much closer parity than we are now, and it seems to me it's better to have this parser be more stringent than plutil, instead of the other way around.

### CODE REVIEW

**Comment by korylprince** [comment]
Note: the original test used `os.ReadDir` and `os.ReadFile`, which failed the testers running Go versions older than 1.16. I force-pushed a version that uses ioutil's equivalents instead to appease the testers.

**Comment by korylprince** [comment]
> Would this catch #3?

Yep, the `testdata/xml/invalid-data.plist` case matches #3.

File: decode_test.go [hunk 1]
```diff
@@ -513,3 +514,52 @@ func TestDecodeTagSkip(t *testing.T) {
 		t.Error("field decoded when it was tagged as -")
 	}
 }
+
+// test parity with plutil -lint on macOS
+var xmlParityTestFailures = map[string]struct{}{
```
Comment (jessepeterson) [inline decode_test.go:519]: nit: I don't like the magic dictionary here. What if you encoded the plist testability within the filename? For example `empty-plist.plist` -> `empty-plist.plutil.plist`?

again, non-blocker. :)

**Review by jessepeterson — APPROVED** [review]

File: decode_test.go (hunk 1 above)
Comment (korylprince) [reply decode_test.go]: @jessepeterson I added a commit with this suggestion. Can you re-review `decode_test.go: TestXMLPlutilParity`?

**Review by jessepeterson — APPROVED** [review]
thanks @korylprince!

**Pull request merged**